File: Day10-19/16.py
def gen_data(o, n):
    while len(o) < n:
        m = ['1' if i == '0' else '0' for i in o[::-1]]
        o += '0' + ''.join(m)
    return o[:n]


def checksum(s):
    c = ''
    for i in range(0, len(s), 2):
        c += '1' if s[i] == s[i+1] else '0'
    return c if len(c) % 2 else checksum(c)

if __name__ == '__main__':
    assert checksum(gen_data('10000', 20)) == '01100'

    print("Part 1: %s " % checksum(gen_data('10111011111001111', 272)))
    print("Part 1: %s " % checksum(gen_data('10111011111001111', 35651584)))

File: Day10-19/16.js
'use strict';

function generate_data(inp, len){
  while(inp.length < len) {
    var mir = inp.split('').reverse().map((x) => {
      return x == '0' ? '1' : '0'
    })
    inp += '0' + mir.join('')
  }
  return inp.slice(0, len);
}

function checksum(cs){
  while(cs.length % 2 == 0) {
    var new_cs = ''
    for(var i = 0; i < cs.length; i += 2) {
      new_cs += cs[i] == cs[i+1] ? '1' : '0'
    }
    cs = new_cs
  }
  return cs
}

console.log(checksum(generate_data('10000', 20)) == '01100')
console.log('Part 1: ' + checksum(generate_data('10111011111001111', 272)))
console.log('Part 1: ' + checksum(generate_data('10111011111001111', 35651584)))
